### DISCUSSION TITLE
Proposal: a persistent Virtual FileSystem

### DISCUSSION THREAD

**Discussion by WebReflection** [pyscript/pyscript#1973] [General]
In more than an occasion our users asked how to persist data in PyScript or how to share the FileSystem between *main* and *workers*.

There are, in my opinion, at least 2 ways to go here:

  * provide an easy to import and use dedicated API (or copy whatever Python FS APIs do) that is based on IndexedDB
  * allow a flag that [mount an IndexedDB based FS](https://pyodide.org/en/stable/usage/file-system.html#mounting-a-file-system) instead of the memory one both projects default to

### Case 1: We provide the API

Possibly the most user-friendly way to go, we could use [this project of mine](https://github.com/WebReflection/fs#readme) to actually expose a `pyfs` module or a `pyscript.fs` module that brings in most common primitives to read, write, crawl, or stat a filesystem. We might discuss if it makes sense to have the API always async or if we want to expose it only in workers and have it fully sync thanks to Atomics.

Workers will incrementally be then the way to provide super-powers to PyScript but arguably, as we do have `async/await` ability in python, even top-level when needed, I feel like we might be good with an always async approach so that on main, one could land MicroPython thanks to its faster bootstrap and on workers either MicroPython or Pyodide and these can share the same persistent storage.

On the other hand, this approach is manual and it might bring some confusion because it won't be fully integrated with the already provided VFS from either Pyodide or MicroPython ...

### Case 2: We override the default mount point

This option should work out of the box (but I haven't tried yet) and it should make any FS related operation persistent on every further run of the same project.

My only concern here is that all interpreters will use a library a part to perform `syncfs()` operations and it might not be easy to reason about the need of async/await between operations, although this is theoretical.

In this scenario we'll have a `persistent_vfs = true` flag that, if enabled, automatically un-mount the MEMFS and mount the IDBFS one and we do call `syncfs()` on writes ... but I wonder if there are possible gotchas I am not considering in here.

@hoodmane @dpgeorge @ntoll @antocuni @fpliger what are your thoughts? Any preference? Any blocker? Any caveat to consider? 🤔

**Comment by fpliger** [comment]
@WebReflection can you elaborate on 

```
My only concern here is that all interpreters will use a library a part to perform syncfs() operations and it might not be easy to reason about the need of async/await between operations, although this is theoretical.
```
?

Wouldn't the underlying interpreters provide that out of the box?

**Reply by WebReflection** [reply]
basically Emscripten offers VFS as MEMFS or IDBFS or other variants ... the current one is always synchronous and *instant*, the IDBFS variant requires `syncfs()` calls to reflect the state or pause before reads cna be atomic/consistent.

If we provide an API we are probably better off by making this fully async: `await read(...)`, `await write(...)`, `await stat(...)`, `await mkdir(...)` ... and so on.

If we just override the MEMFS mount point with an async IDBFS one, most code that work now might fail if users don't change their code too, due to the different nature of the two FS, one is sync, one is async.

Not sure I've explained it any better ... but to add anything, if a worker writes (or main) and then the other part reads without any pause on `syncfs()` the results might surprise while if both sides do `syncfs()` I don't know Emscripten internals enough to be sure that would work if the second call happens while the first one is still going but a new write was queued ... I assume yes, but we need some testing around this.

**Reply by fpliger** [reply]
Thanks, yeah, that makes sense. It's a great topic for the community call tomorrow :)

**Comment by dpgeorge** [comment]
MicroPython has a VFS which is transparent to the user if they don't want to think about it, but also allows mounting arbitrary filesystems (even ones defined in Python).  By default on the `webassembly` port it mounts the Emscripten/POSIX filesystem.  But you can unmount this and then mount whatever you like.

BUT, it's all synchronous, because it has to match the normal CPython API, eg the `os` module.

For microcontrollers an async filesystem would be useful but we haven't made any progress in that direction.

I would suggest that PyScript investigate what an async filesystem would look like, IMO that is more future proof than a synchronous filesystem.

**Reply by WebReflection** [reply]
> BUT, it's all synchronous

you mean ... the code around expects this to be synchronous, right? The Emscripten IDBFS FileSystem requires an inevitably asynchronous operation to persist, namely [FS.syncfs](https://emscripten.org/docs/api_reference/Filesystem-API.html#FS.syncfs).

It is not strictly needed and everything should work out of the box on either Pyodide or MicroPython side **but** if you want that data persistent **and** synchronized with other threads that trust that data, all `write` operations should be followed by that call, hence at least `write()` could be simplified as `await FS.write(buffer)`.

This would automatically persist data on the browser's IndexedDB space and, once persisted, should also be reflected in other threads.

However, last time I've tested, other threads might be out of sync if the data from the DB is not reflected into memory so that actually that `await write(...)` should probably do the following:

  * write data
  * persist via `FS.syncfs(true, callback)`
  * the callback, once resolved without errors, should tell all threads attached to the same IndexedDB that they also need to `FS.sync(true)` in their own realm/environment
  * the initial write can be resolved or rejected

I think today I'll experiment with this and see if I can create a convincing demo for the community call.

There are other topics such as cache-invalidation (modules) if a file already loaded gets some write for some reason but I don't think we need to tackle such edge cases already ... let's start small.

**Reply by WebReflection** [reply]
**Update** ... I've just realized MicroPython doesn't include IDBFS so that `FS.filesystems.IDBFS` is `undefined` and we cannot use it.

**However** I've also failed at using the same `/idbfs` mount point in Pyodide in a persistent manner and multiple pyodide instances seem to be incapable of sharing the same FS ... likely I am doing something wrong but so far I had no luck at all with neither interpreters ... @hoodmane any idea why would that be or if it's even possible?

**Reply by hoodmane** [reply]
I'm not sure about mounting custom file systems at the root directory -- it seems unlikely to me to work. Try mounting in a new directory first.

I'm pretty sure that idbfs wasn't designed to be robust to sharing between multiple instances there will be a bunch of race conditions. You could try to write your own version that's thread safe but it's probably quite tricky.

You might want to check out browserfs:
https://github.com/jvilk/browserfs

**Reply by WebReflection** [reply]
I do have a module for that https://github.com/WebReflection/fs and it's been proven to work already as generic root mount point and to work reliably cross realms (main + worker) so it could work, but after trying with Pyodide to always bootstrap the same folder and read/write to it, always after explicitly awaited sync, the new pyodide can't find anything in there and I start wondering if you have any test page to understand where I am failing at using IDBFS in my code, thanks!

**Reply by WebReflection** [reply]
About that ... one thing comes to mind ... we do create object URLs for workers code and I wonder, at this point, if that might be the culprit: every runtime created object URL won't share the same IndexedDB as the main thread that created such object URL ... I will investigate this and provide answers and, if that's a browsers' issue, I am afraid we can't move forward with this idea as that would not work by standards meaning of objects URLs 😢

**edit** that being said, my tests with Pyodide were both on the main thread so something more to understand here to me.

**Reply by james-pre** [reply]
Hi @hoodmane, I'm the maintainer of BrowserFS. I'd like to let you know that BrowserFS has been deprecated. You can read more about that on the BrowserFS readme (https://github.com/jvilk/BrowserFS). [ZenFS](https://github.com/zen-fs/core) is an updated and active fork that I created.

**Comment by ntoll** [comment]
I've been thinking a lot about this. I'd love some clarifications/comments:

* What problem are we trying to solve here? Put simply I think it can be summarised simply as "persistent shared storage".
* Is a file system the correct abstraction to use for the solution to this problem? Previous experience (my last gig where users had a shared `/home/{username}` drive between different VMs) tells me that there is huge potential for mess. I'd like to echo @hoodmane's point about race conditions and add my own: we'd be giving our users a shotgun and, due to subtle and often unforeseen consequences, they'll regularly shoot themselves in the foot with it. :boom: This is likely to create an unwanted support burden for us. Speaking from experience, such FS related support is fiddly, subtle and often ends with, "just don't do that, ok?". :wink: 
* Abstracting a FS feels a very "heavy" solution (we need to write, debug and maintain new code). Code that doesn't exist can't have bugs in it. :+1:
* So, is there are better abstraction we can use for persistent shared storage in PyScript? Is there some notion of an ACID like data store in the browser realm around which we can create a simple Pythonic shim to achieve the same ends?

These questions are offered in a spirit of technical exploration. As always, remember I have a Winnie-the-Pooh small brain so likely missing something obvious. :grin: 

As always, comments, suggestions, constructive critique and ideas are most welcome. :rocket:

**Reply by WebReflection** [reply]
I can answer most questions:

  * users have been asking about persistent data across sessions ... so you refresh, data is already there ... right now every refresh or session starts from nothing and `localStorage` or even `sessionStorage` are a (very limited and blocking) JS affair, not really Python
  * the `/home/{username}` makes little sense here as the persistent storage is **per browser** (*and* per domain), not stored in the server ... if you want to create `/home/{username}` you can already do that with the current VFS in memory but it needs to be created every single time you go back to that page ... we can't prevent users to mess with the VFS but most importantly, IndexedDB has no race conditions issues, it's ATOMIC within its transactions and the only race is if you read data while another thread is writing data ... in such case you won't get the latest/greatest but we have *SharedWorker* to the resque, eventually, to guarantee that there is a single orchestration of the read/write cycle and we can synchronize in there all the things
  * if we keep it minimal, stat, read, write, dir are all we need but I agree we should not reinvent the wheel, specially with a PL such as Python that already has its own FileSystem primitives ... that's why I'd like to be integrated eventually with the VFS in Pyodide or MicroPython so that they just use everything they use already but that would persist as IDBFS storage type, which is allowed by Pyodide but it doesn't work out of the box (it's absent in MicroPython but [this is just Emscripten providing it](https://github.com/WebReflection/fs#readme), nothing to maintain from our side and nothing different from the current in memory VFS we use already).
  * AFAIK IndexedDB is ACID and so is SQLite ... [I've combined both already](https://github.com/WebReflection/sqlite-worker#readme) and it works great but it shares nothing across multiple instantiations ... as both APIs could be used in a SharedWorker and SQLite is very well documented and also already available in Python, all we could do is create a SharedWorker and provide a `shared_sqlite` module (not in stdlib) that could allow both Pyodide and MicroPython to share the same data per domain and browser, as it is already for IndexedDB and SharedWorker and dictated by Web standards security concerns, and offer an async SQLite API with ease ... without abstracting, just bootstrapping the right thing in the right way. This could be doable, but I am not sure if it fulfills current users requirements ... although having it always `async` removes the *SharedArrayBuffer* headers burden from the equation, working with or without those headers around.

**edit** I did write SharedArrayBuffer instead of [SharedWorker](https://developer.mozilla.org/en-US/docs/Web/API/SharedWorker), now corrected.
